Alerting: Add error to annotations when execution error state is Alerting

alexander-akhmetov opened the pull request:
**What is this feature?**

Add error to annotations when execution error state is Alerting

**Why do we need this feature?**

Currently we do not add anything to the annotations, so when the rule is configured to be `Alerting` on errors, there is no additional information about the error in the alert notifications.

**Without the change**
<img width="658" alt="image" src="https://github.com/user-attachments/assets/4dab0732-5584-447a-be96-c7a2ff56397a">


**With the change**
New `Error`, `datasource_uid` and `ref_id` annotations.
<img width="690" alt="image" src="https://github.com/user-attachments/assets/bec030aa-75ab-4edd-82a4-fca124ec3333">


<details>
  <summary><b>Grafana UI</b></summary>

### Alerting

<img width="1471" alt="image" src="https://github.com/user-attachments/assets/e9bc3729-e9be-4199-979f-133ca1dc5a46">

### Back to normal

<img width="1466" alt="image" src="https://github.com/user-attachments/assets/1c5cfa8b-06ef-42f7-ab4f-966b0ae13a9c">


</details>

alexander-akhmetov committed:
Alerting: Add error to annotations when execution error state is Alerting

alexander-akhmetov commented on pkg/services/ngalert/state/manager.go:
```diff
@@ -463,7 +463,7 @@ func (st *Manager) setNextState(ctx context.Context, alertRule *ngModels.AlertRu
 	// This is temporary change to make sure that the labels are not persistent in the state after it was in Error state
 	// TODO yuri. Remove it when correct Error result with labels is provided
-	if currentState.State == eval.Error && result.State != eval.Error {
+	if (currentState.State == eval.Error || currentState.State == eval.Alerting) && result.State != eval.Error {
```
Maybe with `currentState.State == eval.Alerting` we should also check that `rule.ExecErrState == models.AlertingErrState`? Not sure

alexander-akhmetov committed:
Merge branch 'main' of github.com:grafana/grafana into alexander-akhmetov/add-error-to-annotations-when-alerting

yuri-tceretian commented on pkg/services/ngalert/state/manager_private_test.go:
```diff
@@ -3689,14 +3689,17 @@ func TestProcessEvalResults_StateTransitions(t *testing.T) {
 			},
 			{
-				desc: "t1[QueryError] t2[{}:normal] at t2",
+				desc: "t1[QueryError] t2[{}:normal] at t2,t3",
```
please change description to "t1[QueryError] t2[{}:normal] t3[{}:normal] at t2,t3",

alexander-akhmetov replied:
updated!

yuri-tceretian reviewed (changes requested):
when Error is mapped to either Alerting or Normal, no new labels should be added to the state. Otherwise, we break notification pipline. `ref_id` and `datasource_uid` should be included into annotations.

alexander-akhmetov committed:
Add datasource_uid and ref_id to annotations

alexander-akhmetov commented:
Updated to use annotations: https://github.com/grafana/grafana/pull/91594/commits/026178ba9310ef9731867fcf2ba6253b9bfedba2

alexander-akhmetov committed:
Add Error annotations to the normal state on datasource errors

When an alert rule is configured to remain in the Normal state on datasource errors,
Grafana now adds annotations to the alert instance. These annotations provide
details about the error and the associated datasource.

alexander-akhmetov committed:
Add Error annotations to the KeepLastErrState state on datasource errors

yuri-tceretian reviewed (approved):
LGTM! Great change!
